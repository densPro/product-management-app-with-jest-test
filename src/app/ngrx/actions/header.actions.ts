import { createAction, props } from '@ngrx/store';

export const updateHeaderTitle = createAction(
  '[Header] Update Title',
  props<{ title: string }>()
);