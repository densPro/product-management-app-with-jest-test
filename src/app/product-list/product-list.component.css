.product-card {
  max-width: 400px; 
  padding: 20px;
  margin: 10px;
}

.product-list {
  display: flex;
  flex-wrap: wrap;
  padding: 20px;
  justify-content: center;
}

.action-buttons {
  justify-content: flex-end;
}

.add-button {
  position: fixed;
  bottom: 20px; /* Adjust as needed */
  right: 20px; /* Adjust as needed */
  z-index: 1000; /* Ensure it's above other content */
}