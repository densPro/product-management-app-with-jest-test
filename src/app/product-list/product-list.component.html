<div class="product-list">
  <mat-card class="product-card" *ngFor="let product of products$ | async">
    <mat-card-header>
      <mat-card-title>{{ product.name }}</mat-card-title>
    </mat-card-header>
    <mat-card-content>
      <p>{{ product.description }}</p>
      <mat-card-subtitle>${{ product.price }}</mat-card-subtitle>
    </mat-card-content>
    <mat-card-actions class="action-buttons">
      <button mat-icon-button (click)="editProduct(product.id)">
        <mat-icon>edit</mat-icon>
      </button>
    </mat-card-actions>
  </mat-card>

  <button mat-fab color="primary" class="add-button" (click)="addProduct()">
    <mat-icon>add</mat-icon>
  </button>
</div>
