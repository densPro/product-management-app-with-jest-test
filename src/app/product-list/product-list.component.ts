import { Component } from '@angular/core';
import { CommonModule } from '@angular/common';
import { Observable, map } from 'rxjs';
import { MatButtonModule } from '@angular/material/button';
import { MatCardModule } from '@angular/material/card';
import { MatIconModule } from '@angular/material/icon';
import { Router } from '@angular/router';
import { Store } from '@ngrx/store';

import { Product } from '../product.model';
import * as fromActions from '../ngrx/actions/header.actions';
import { ProductService } from '../product.service';

@Component({
  selector: 'app-product-list',
  standalone: true,
  imports: [CommonModule, MatCardModule, MatButtonModule, MatIconModule],
  templateUrl: './product-list.component.html',
  styleUrl: './product-list.component.css',
})
export class ProductListComponent {
  products$: Observable<Product[]> | undefined;

  constructor(
    private productService: ProductService,
    private router: Router,
    private store: Store
  ) {
    this.store.dispatch(
      fromActions.updateHeaderTitle({ title: 'Products' })
    );
  }

  ngOnInit() {
    this.products$ = this.productService
      .getProducts()
      .pipe(map((products) => products.slice().sort((a, b) => b.id - a.id)));
  }

  editProduct(id: number) {
    this.router.navigateByUrl(`/products/${id}`);
  }

  addProduct() {
    this.router.navigateByUrl(`/add-product`);
  }
}
