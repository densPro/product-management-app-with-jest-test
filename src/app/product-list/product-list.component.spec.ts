import { ComponentFixture, TestBed } from '@angular/core/testing';
import { ProductListComponent } from './product-list.component';
import { CommonModule } from '@angular/common';
import { MatCardModule } from '@angular/material/card';
import { MatButtonModule } from '@angular/material/button';
import { MatIconModule } from '@angular/material/icon';
import { Observable, of } from 'rxjs';
import { Product } from '../product.model';
import { Router } from '@angular/router';
import { Store } from '@ngrx/store';
import { MockStore } from '@ngrx/store/testing';
import { ProductService } from '../product.service';
import * as fromActions from '../ngrx/actions/header.actions';

describe('ProductListComponent', () => {
  let component: ProductListComponent;
  let fixture: ComponentFixture<ProductListComponent>;
  let router: Router;
  let store: MockStore;
  let productServiceSpy: jasmine.SpyObj<ProductService>;
  const mockProducts: Product[] = [
    { id: 1, name: 'Product 1', description: 'Description 1', price: 10 },
    { id: 2, name: 'Product 2', description: 'Description 2', price: 20 },
  ];

  beforeEach(async () => {
    productServiceSpy = jasmine.createSpyObj('ProductService', ['getProducts']);

    await TestBed.configureTestingModule({
      imports: [
        ProductListComponent,
        CommonModule,
        MatCardModule,
        MatButtonModule,
        MatIconModule,
      ],
      providers: [
        { provide: Router, useValue: { navigateByUrl: jasmine.createSpy() } }, 
        { provide: Store, useValue: { dispatch: jasmine.createSpy() } }, 
        { provide: ProductService, useValue: productServiceSpy }, 
      ],
    }).compileComponents();

    router = TestBed.inject(Router);
    store = TestBed.inject(Store) as MockStore;
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(ProductListComponent);
    component = fixture.componentInstance;
    productServiceSpy.getProducts.and.returnValue(of(mockProducts));
    fixture.detectChanges();
  });

  it('should create the component', () => {
    expect(component).toBeTruthy();
  });

  it('should load products on initialization', () => {
    expect(component.products$).toBeDefined();
    expect(component.products$ instanceof Observable).toBe(true);
    component.products$!.subscribe(products => {
      expect(products.length).toEqual(mockProducts.length);
      const sortedProducts = products.sort((a, b) => a.id - b.id);
      sortedProducts.forEach((product, index) => {
        expect(product.id).toEqual(mockProducts[index].id);
        expect(product.name).toEqual(mockProducts[index].name);
        expect(product.description).toEqual(mockProducts[index].description);
        expect(product.price).toEqual(mockProducts[index].price);
      });
    });
  });

  it('should dispatch updateHeaderTitle action with correct title', () => {
    expect(store.dispatch).toHaveBeenCalledWith(fromActions.updateHeaderTitle({ title: 'Products' }));
  });

  it('should navigate to edit product route when editProduct is called', () => {
    const productId = 1;
    component.editProduct(productId);
    expect(router.navigateByUrl).toHaveBeenCalledWith(`/products/${productId}`);
  });

  it('should navigate to add product route when addProduct is called', () => {
    component.addProduct();
    expect(router.navigateByUrl).toHaveBeenCalledWith('/add-product');
  });
});
