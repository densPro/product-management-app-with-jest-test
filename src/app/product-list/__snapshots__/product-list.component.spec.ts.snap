// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductListComponent should match snapshot 1`] = `
<div
  id="root1"
>
  <div
    class="product-list"
  >
    <mat-card
      class="mat-mdc-card mdc-card product-card"
    >
      <mat-card-header
        class="mat-mdc-card-header"
      >
        <div
          class="mat-mdc-card-header-text"
        >
          <mat-card-title
            class="mat-mdc-card-title"
          >
            Product 2
          </mat-card-title>
        </div>
      </mat-card-header>
      <mat-card-content
        class="mat-mdc-card-content"
      >
        <p>
          Description 2
        </p>
        <mat-card-subtitle
          class="mat-mdc-card-subtitle"
        >
          $20
        </mat-card-subtitle>
      </mat-card-content>
      <mat-card-actions
        class="mat-mdc-card-actions mdc-card__actions action-buttons"
      >
        <button
          class="mdc-icon-button mat-mdc-icon-button mat-unthemed mat-mdc-button-base"
          mat-icon-button=""
          mat-ripple-loader-centered=""
          mat-ripple-loader-class-name="mat-mdc-button-ripple"
          mat-ripple-loader-uninitialized=""
        >
          <span
            class="mat-mdc-button-persistent-ripple mdc-icon-button__ripple"
          />
          <mat-icon
            aria-hidden="true"
            class="mat-icon notranslate material-icons mat-ligature-font mat-icon-no-color"
            data-mat-icon-type="font"
            role="img"
          >
            edit
          </mat-icon>
          <span
            class="mat-mdc-focus-indicator"
          />
          <span
            class="mat-mdc-button-touch-target"
          />
        </button>
      </mat-card-actions>
    </mat-card>
    <mat-card
      class="mat-mdc-card mdc-card product-card"
    >
      <mat-card-header
        class="mat-mdc-card-header"
      >
        <div
          class="mat-mdc-card-header-text"
        >
          <mat-card-title
            class="mat-mdc-card-title"
          >
            Product 1
          </mat-card-title>
        </div>
      </mat-card-header>
      <mat-card-content
        class="mat-mdc-card-content"
      >
        <p>
          Description 1
        </p>
        <mat-card-subtitle
          class="mat-mdc-card-subtitle"
        >
          $10
        </mat-card-subtitle>
      </mat-card-content>
      <mat-card-actions
        class="mat-mdc-card-actions mdc-card__actions action-buttons"
      >
        <button
          class="mdc-icon-button mat-mdc-icon-button mat-unthemed mat-mdc-button-base"
          mat-icon-button=""
          mat-ripple-loader-centered=""
          mat-ripple-loader-class-name="mat-mdc-button-ripple"
          mat-ripple-loader-uninitialized=""
        >
          <span
            class="mat-mdc-button-persistent-ripple mdc-icon-button__ripple"
          />
          <mat-icon
            aria-hidden="true"
            class="mat-icon notranslate material-icons mat-ligature-font mat-icon-no-color"
            data-mat-icon-type="font"
            role="img"
          >
            edit
          </mat-icon>
          <span
            class="mat-mdc-focus-indicator"
          />
          <span
            class="mat-mdc-button-touch-target"
          />
        </button>
      </mat-card-actions>
    </mat-card>
    
    <button
      class="add-button mdc-fab mat-mdc-fab mat-primary mat-mdc-button-base"
      color="primary"
      mat-fab=""
      mat-ripple-loader-class-name="mat-mdc-button-ripple"
      mat-ripple-loader-uninitialized=""
    >
      <span
        class="mat-mdc-button-persistent-ripple mdc-fab__ripple"
      />
      <mat-icon
        aria-hidden="true"
        class="mat-icon notranslate material-icons mat-ligature-font mat-icon-no-color"
        data-mat-icon-type="font"
        role="img"
      >
        add
      </mat-icon>
      <span
        class="mdc-button__label"
      />
      <span
        class="mat-mdc-focus-indicator"
      />
      <span
        class="mat-mdc-button-touch-target"
      />
    </button>
  </div>
</div>
`;
